src/handleInteraction.ts

```diff
@@ -0,0 +1,192 @@
+	}
+
+	const context: CommandContext = {
+		createdTimestamp: interaction.createdTimestamp,
+		user: interaction.user,
+		member,
+		guild: interaction.guild,
+		channel,
+		client: interaction.client,
+		interaction,
+		options: interaction.options.data,
+		logger,
+		async prepareForLongRunningTasks(ephemeral) {
+			try {
+				await interaction.deferReply({ ephemeral });
+			} catch (error) {
+				logger.error('Failed to defer reply to interaction:', error);
+			}
+		},
+		async replyPrivately(options, viaDM: boolean = false) {
+			if (viaDM) {
+				// We need to say *something* to the interaction itself, or Discord will think we died.
+				const content = ':paperclip: Check your DMs';
+				if (interaction.deferred) {
+					try {
+						await interaction.editReply(content);
+					} catch (error) {
+						logger.error('Failed to edit reply to interaction:', error);
+					}
+				} else {
+					try {
+						await interaction.reply({ content, ephemeral: true });
+					} catch (error) {
+						logger.error('Failed to reply to interaction:', error);
+					}
+				}
+			}
+			if (interaction.deferred && !viaDM) {
+				try {
+					if (typeof options === 'string') {
+						await interaction.followUp({ ephemeral: true, content: options });
+					} else {
+						await interaction.followUp({ ...options, ephemeral: true });
+					}
+				} catch (error) {
+					logger.error('Failed to follow up on interaction:', error);
+				}
+			} else {
+				const reply = await replyPrivately(interaction, options, viaDM);
+				if (reply === false) {
+					logger.info(`User ${logUser(interaction.user)} has DMs turned off.`);
+				}
+			}
+		},
+		async reply(options) {
+			if (interaction.deferred) {
+				try {
+					await interaction.editReply(options);
+				} catch (error) {
+					logger.error('Failed to edit reply to interaction:', error);
+					await interaction.followUp(options);
+				}
+			} else {
+				try {
+					if (typeof options === 'string') {
+						await interaction.reply(options);
+					} else if (
+						!('shouldMention' in options) ||
+						options.shouldMention === undefined ||
+						options.shouldMention
+					) {
+						// Doesn't say whether to mention, default to `true`
+						await interaction.reply(options);
+					} else {
+						// Really shouldn't mention
+						await interaction.reply({
+							...options,
+							allowedMentions: { users: [] },
+						});
+					}
+				} catch (error) {
+					logger.error('Failed to reply to interaction:', error);
+				}
+			}
+
+			if (typeof options !== 'string' && 'ephemeral' in options && options?.ephemeral === true) {
+				// FIXME: We didn't actually send the reply if we errored out
+				logger.info(
+					`Sent ephemeral reply to User ${logUser(interaction.user)}: ${JSON.stringify(options)}`
+				);
+			}
+		},
+		async followUp(options) {
+			if (
+				typeof options !== 'string' &&
+				(!('reply' in options) || options.reply === false || options.reply === undefined) &&
```

Comment:
The `'reply' in options` bit tells TypeScript that it's safe to access `options.reply`. Since `options` can be neither `null` or `undefined`, the `?.` operator is useless to us here.

As for why we couldn't just use `!options.reply` there, that's due to an ESLint configuration I have checking for nullable values in conditionals, to make those checks explicit. Sometimes, we may want different behavior for nullish values vs empty or falsy ones, and I don't always keep the distinction in my mind when I'm writing conditions, hence the ESLint rule enforcing explicit logic.

Reply:
sounds good